### PR TITLE
Content assist does not activate, inside if() within a switch

### LINKED ISSUE #1561: Content assist does not activate, inside if() within a switch
_I'm migrating [bugzilla 576934](https://bugs.eclipse.org/bugs/show_bug.cgi?id=576934) here, as requested in (the closely-related) github issue #341_

I had upgraded from Eclipse 4.20 > 4.21 to pick up fixes for content assist regressions, but I ran into this still-failing case:

    public class Snippet {
      final String name = "test";
      enum Type { A }

      private void go(Type a) {
        switch (a) {
        case A:
          if (nam)
          break;
        }
      }
    }

Trying ctrl+space after 'nam' in the incomplete 'if' statement still shows nothing, similarly to [bugzilla 573632](https://bugs.eclipse.org/bugs/show_bug.cgi?id=573632)

Also per a workaround mentioned in there, adding another symbol after the identifier gets completion working again:

  `if (nam <)`

Deleting the closing ')' also kicks it into working:

  `if (nam`

Note, however, that Eclipse by default inserts that closing ')' for us. 

I have confirmed this is still present in a recent nightly - Eclipse 2023-12 (4.30) Build id: I20231024-1800

### PR DESCRIPTION
**Pull request by stephan-herrmann**
fixes #1561

Simpler approach, using new tests from #1625 

@gayanper I tried to explain with minimal code comments. LMK if this is sufficient to understand the approach.

This change is not quite complete yet:
* no explanation for changed relevance in three tests
* may need to cover more control statements: only implemented for those that are covered by tests from [#1625](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1625)

### CODE REVIEW

**Comment by gayanper** [comment]
@stephan-herrmann This approach is much elegant. I think we should go ahead with this.

File: org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/model/CompletionTests18.java [hunk 1]
```diff
@@ -5951,7 +5951,7 @@ public void testBug574823_completeOn_methodInvocationWithParams_inIfConidtion_in
 
 	String result = requestor.getResults();
     assertTrue(String.format("Result doesn't contain method forEach (%s)", result),
-    		result.contains("forEach[METHOD_REF]{forEach(), Ljava.lang.Iterable<Ljava.lang.String;>;, (Ljava.util.function.Consumer<-Ljava.lang.String;>;)V, null, null, forEach, (arg0), replace[149, 149], token[149, 149], 60}"));
+    		result.contains("forEach[METHOD_REF]{forEach(), Ljava.lang.Iterable<Ljava.lang.String;>;, (Ljava.util.function.Consumer<-Ljava.lang.String;>;)V, null, null, forEach, (arg0), replace[149, 149], token[149, 149], 55}"));
```
Comment (gayanper) [inline org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/model/CompletionTests18.java:5954]: @stephan-herrmann  Could this be due to poping out the `interestingEnclosings` ?

**Comment by gayanper** [comment]
I tried to see if we are missing any control conditions, I see we don't have a endVisit for ForStatement, But even without it our unit test pass with following test snippet

```
switch (input) {
case A:
	for (int i = 0; i < nam)
	break;
}
```
Do you think we are missing any ?

File: org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/model/CompletionTests18.java (hunk 1 above)
Comment (stephan-herrmann) [reply org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/model/CompletionTests18.java:5954]: > @stephan-herrmann Could this be due to poping out the `interestingEnclosings` ?

To assess the change in relevance I'd first check which ingredient to this number got lost.

**Comment by stephan-herrmann** [comment]
> I tried to see if we are missing any control conditions, I see we don't have a endVisit for ForStatement, But even without it our unit test pass with following test snippet
> 
> ```
> switch (input) {
> case A:
> 	for (int i = 0; i < nam)
> 	break;
> }
> ```
> 
> Do you think we are missing any ?

To parallel the original test case I'd try smth like this:
```switch (input) {
case A:
	for (int i = 0; nam|)
	    break;
}
```
i.e., completion on a _direct_ child in the ForStatement.condition location (currently you have an intermediate binary expression).

Another check for completeness: if we have a method in scope that returns boolean and has a name starting with 'nam': do we propose it with relevance `R_EXACT_EXPECTED_TYPE`? I.e. is completion correctly leveraging the information from the .condition location? I'm not sure though, if CompletionEngine is complete in this regard, yet.

**Comment by gayanper** [comment]
@stephan-herrmann I added a additional test for testing the scenarios you mentioned in above comment https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1953#issuecomment-1937844419

**Pull request merged**